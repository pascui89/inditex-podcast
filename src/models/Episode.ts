export interface Episode {
  episodeUrl: string;
  trackTimeMillis: number;
  artistIds: number[];
  collectionViewUrl: string;
  description: string;
  shortDescription: string;
  feedUrl: string;
  closedCaptioning: string;
  trackId: number;
  trackName: string;
  releaseDate: string;
  collectionId: number;
  collectionName: string;
  country: string;
  artworkUrl600: string;
  episodeFileExtension: string;
  artworkUrl160: string;
  episodeContentType: string;
  previewUrl: string;
  artworkUrl60: string;
  artistViewUrl: string;
  trackViewUrl: string;
  genres: { name: string; id: string }[];
  episodeGuid: string;
  kind: string;
  wrapperType: string;
}
