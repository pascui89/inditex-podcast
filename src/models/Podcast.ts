export interface Podcast {
  feed: {
    author: {
      name: {
        label: string;
      };
      uri: {
        label: string;
      };
    };
    entry: PodcastEntry[];
    updated: {
      label: string;
    };
    rights: {
      label: string;
    };
    title: {
      label: string;
    };
    icon: {
      label: string;
    };
    link: {
      attributes: {
        rel: string;
        type: string;
        href: string;
      };
    }[];
    id: {
      label: string;
    };
  };
}

export interface PodcastEntry {
  'im:name': {
    label: string;
  };
  'im:image': {
    label: string;
    attributes: {
      height: string;
    };
  }[];
  summary: {
    label: string;
  };
  'im:price': {
    label: string;
    attributes: {
      amount: string;
      currency: string;
    };
  };
  'im:contentType': {
    attributes: {
      term: string;
      label: string;
    };
  };
  rights: {
    label: string;
  };
  title: {
    label: string;
  };
  link: {
    attributes: {
      rel: string;
      type: string;
      href: string;
    };
  };
  id: {
    label: string;
    attributes: {
      'im:id': string;
    };
  };
  'im:artist': {
    label: string;
    attributes?: {
      href: string;
    };
  };
  category: {
    attributes: {
      'im:id': string;
      term: string;
      scheme: string;
      label: string;
    };
  };
  'im:releaseDate': {
    label: string;
    attributes: {
      label: string;
    };
  };
}

export const podcastMock: Podcast = {
  feed: {
    author: {
      name: {
        label: 'Author Name',
      },
      uri: {
        label: 'Author URI',
      },
    },
    entry: [
      {
        'im:name': {
          label: 'Podcast Name 1',
        },
        'im:image': [
          {
            label: 'Image URL 1',
            attributes: {
              height: '100',
            },
          },
          {
            label: 'Image URL 1',
            attributes: {
              height: '100',
            },
          },
          {
            label: 'Image URL 1',
            attributes: {
              height: '100',
            },
          },
        ],
        summary: {
          label: 'Podcast Summary 1',
        },
        'im:price': {
          label: '9.99',
          attributes: {
            amount: '9.99',
            currency: 'USD',
          },
        },
        'im:contentType': {
          attributes: {
            term: 'Term 1',
            label: 'Label 1',
          },
        },
        rights: {
          label: 'Podcast Rights 1',
        },
        title: {
          label: 'Podcast Title 1',
        },
        link: {
          attributes: {
            rel: 'Rel 1',
            type: 'Type 1',
            href: 'Link URL 1',
          },
        },
        id: {
          label: 'Podcast ID 1',
          attributes: {
            'im:id': '1',
          },
        },
        'im:artist': {
          label: 'Artist 1',
          attributes: {
            href: 'Artist URL 1',
          },
        },
        category: {
          attributes: {
            'im:id': '1',
            term: 'Category Term 1',
            scheme: 'Category Scheme 1',
            label: 'Category Label 1',
          },
        },
        'im:releaseDate': {
          label: 'Release Date 1',
          attributes: {
            label: 'Release Date Label 1',
          },
        },
      },
    ],
    updated: {
      label: 'Updated Label',
    },
    rights: {
      label: 'Feed Rights',
    },
    title: {
      label: 'Feed Title',
    },
    icon: {
      label: 'Feed Icon',
    },
    link: [
      {
        attributes: {
          rel: 'Rel 1',
          type: 'Type 1',
          href: 'Link URL 1',
        },
      },
    ],
    id: {
      label: 'Feed ID',
    },
  },
};
