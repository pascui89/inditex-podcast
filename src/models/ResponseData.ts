import { Episode, Podcast } from '.';

export interface ResponseData {
  resultCount: number;
  results: (Podcast | Episode)[];
}
