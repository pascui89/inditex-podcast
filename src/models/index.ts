export type { Episode } from './Episode';
export type { Podcast, PodcastEntry } from './Podcast';
export type { PodcastDetail } from './PodcastDetail';
export type { ResponseData } from './ResponseData';
