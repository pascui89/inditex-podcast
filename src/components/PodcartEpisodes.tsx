import { Fragment, useEffect } from 'react';
import {
  Paper,
  Skeleton,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow
} from '@mui/material';
import { Link, useParams } from 'react-router-dom';
import { Episode, PodcastDetail } from '../models';
import { millisecondsToMinutesSeconds, toDateFormat } from '../utils';
import { StyledCardContentPaddingNone, StyledCardOverflowAuto, StyledTableCell, StyledTableRow } from './styled';

interface IProp {
  loading: boolean;
  podCastDetail: PodcastDetail | null;
}

/**
 * Component for displaying a list of podcast cards.
 *
 * @param {boolean} loading - Indicates if the component is in a loading state.
 * @param {PodcastDetail | null} podCastDetail - The detailed information about the podcast.
 */
export default function PodcartEpisodes(props: IProp) {
  const { podcastId } = useParams();
  const { podCastDetail, loading } = props;

  useEffect(() => {}, []);

  const renderGridTable = (
    <Fragment>
      <TableContainer component={Paper}>
        <Table className='w-full' aria-label="customized table">
          <TableHead>
            <TableRow>
              <TableCell>Title</TableCell>
              <TableCell align="center">Date</TableCell>
              <TableCell align="center">Duration</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {loading &&
              Array.from({ length: 10 }, (_, index) => (
                <StyledTableRow key={index}>
                  <StyledTableCell
                    component="th"
                    data-testid="skeleton"
                    scope="row"
                    colSpan={3}
                  >
                    <Skeleton
                      key={`skeleton-${index}`}
                      variant="rectangular"
                      width="100%"
                      height={53}
                    />
                  </StyledTableCell>
                </StyledTableRow>
              ))}
            {!loading &&
              podCastDetail?.results.map(
                (podCastDetail: Episode, index: number) => (
                  <StyledTableRow key={index}>
                    <StyledTableCell component="th" scope="row">
                      <Link
                        to={`/podcast/${podcastId}/episode/${podCastDetail.trackId}`}
                      >
                        {podCastDetail.trackName}
                      </Link>
                    </StyledTableCell>
                    <StyledTableCell align="center">
                      {toDateFormat(podCastDetail.releaseDate)}
                    </StyledTableCell>
                    <StyledTableCell align="center">
                      {millisecondsToMinutesSeconds(
                        podCastDetail.trackTimeMillis
                      )}
                    </StyledTableCell>
                  </StyledTableRow>
                )
              )}
          </TableBody>
        </Table>
      </TableContainer>
    </Fragment>
  );

  return (
    <StyledCardOverflowAuto>
      <StyledCardContentPaddingNone>
        <StyledCardContentPaddingNone>
          {renderGridTable}
        </StyledCardContentPaddingNone>
      </StyledCardContentPaddingNone>
    </StyledCardOverflowAuto>
  );
}
