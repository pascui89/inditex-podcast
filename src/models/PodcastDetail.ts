import { Episode } from './Episode';

export interface PodcastDetail {
  resultCount: number;
  results: Episode[];
}

export const mockPodcastDetail: PodcastDetail = {
  resultCount: 21,
  results: [
    {
      country: 'USA',
      artworkUrl600:
        'https://is3-ssl.mzstatic.com/image/thumb/Podcasts126/v4/31/d3/f2/31d3f28f-434c-45cc-17cf-ffcdfda0d78a/mza_9731489490781341652.jpg/600x600bb.jpg',
      feedUrl:
        'https://www.omnycontent.com/d/playlist/e73c998e-6e60-432f-8610-ae210140c5b1/9ff2dac3-12fd-4561-b512-ae33005f64f5/df6d181a-09ea-4bf2-adc6-ae33005f650d/podcast.rss',
      closedCaptioning: 'none',
      collectionId: 1096830182,
      collectionName: 'Drink Champs',
      artworkUrl60:
        'https://is3-ssl.mzstatic.com/image/thumb/Podcasts126/v4/31/d3/f2/31d3f28f-434c-45cc-17cf-ffcdfda0d78a/mza_9731489490781341652.jpg/60x60bb.jpg',
      artistViewUrl:
        'https://itunes.apple.com/us/artist/interval-presents/284341002?mt=2&uo=4',
      artworkUrl160:
        'https://is3-ssl.mzstatic.com/image/thumb/Podcasts126/v4/31/d3/f2/31d3f28f-434c-45cc-17cf-ffcdfda0d78a/mza_9731489490781341652.jpg/160x160bb.jpg',
      episodeContentType: 'audio',
      collectionViewUrl:
        'https://itunes.apple.com/us/podcast/drink-champs/id1096830182?mt=2&uo=4',
      trackTimeMillis: 9813000,
      episodeFileExtension: 'mp3',
      genres: [{ name: 'Music', id: '1310' }],
      episodeGuid: 'e7668686-4c6c-4176-90ba-b01b0112eaf5',
      description: 'N.O.R.E. & DJ EFN are the Drink Champs...',
      shortDescription: '',
      releaseDate: '2023-06-09T07:01:00Z',
      previewUrl:
        'https://chrt.fm/track/BE7515/traffic.omny.fm/d/clips/e73c998e-6e60-432f-8610-ae210140c5b1/9ff2dac3-12fd-4561-b512-ae33005f64f5/e7668686-4c6c-4176-90ba-b01b0112eaf5/audio.mp3?utm_source=Podcast&in_playlist=df6d181a-09ea-4bf2-adc6-ae33005f650d',
      artistIds: [284341002],
      trackViewUrl:
        'https://podcasts.apple.com/us/podcast/episode-367-w-de-la-soul/id1096830182?i=1000616244153&uo=4',
      episodeUrl:
        'https://chrt.fm/track/BE7515/traffic.omny.fm/d/clips/e73c998e-6e60-432f-8610-ae210140c5b1/9ff2dac3-12fd-4561-b512-ae33005f64f5/e7668686-4c6c-4176-90ba-b01b0112eaf5/audio.mp3?utm_source=Podcast&in_playlist=df6d181a-09ea-4bf2-adc6-ae33005f650d',
      trackId: 1000616244153,
      trackName: 'Episode 367 w/ De La Soul',
      kind: 'podcast-episode',
      wrapperType: 'podcastEpisode',
    },
  ],
};
