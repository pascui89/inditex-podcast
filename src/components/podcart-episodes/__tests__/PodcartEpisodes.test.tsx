import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import PodcartEpisodes from '../PodcartEpisodes';
import { mockPodcastDetail } from '../../../models/PodcastDetail';
import '@testing-library/jest-dom/extend-expect';

global.React = React;

describe('PodcartEpisodes', () => {
  it('should renders episode details', () => {
    const { container } = render(
      <Router>
        <PodcartEpisodes loading={false} podCastDetail={mockPodcastDetail} />
      </Router>
    );

    expect(container).toMatchSnapshot();
  });

  it('should renders loading state when loading is true', () => {
    const wrapper = render(
      <Router>
        <PodcartEpisodes loading={true} podCastDetail={null} />
      </Router>
    );

    const skeletonElements = wrapper.getAllByTestId('skeleton');
    expect(skeletonElements.length).toBe(10);
  });

  it('should renders episode details when loading is false and podCastDetail is provided', () => {
    const wrapper = render(
      <Router>
        <PodcartEpisodes loading={false} podCastDetail={mockPodcastDetail} />
      </Router>
    );

    const episodeNameElements = wrapper.getAllByRole('link');
    expect(episodeNameElements.length).toBe(mockPodcastDetail.results.length);
  });
});
