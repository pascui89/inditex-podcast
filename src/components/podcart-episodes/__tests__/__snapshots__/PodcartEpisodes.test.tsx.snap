// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PodcartEpisodes should renders episode details 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-yq514b-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-1aqrhzj-MuiCardContent-root"
    >
      <div
        class="MuiCardContent-root css-1aqrhzj-MuiCardContent-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
        >
          <table
            aria-label="customized table"
            class="MuiTable-root w-full css-rqglhn-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Title
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                  scope="col"
                >
                  Date
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                  scope="col"
                >
                  Duration
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            >
              <tr
                class="MuiTableRow-root css-1enotpu-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1xkql70-MuiTableCell-root"
                  scope="row"
                >
                  <a
                    href="/podcast/undefined/episode/1000616244153"
                  >
                    Episode 367 w/ De La Soul
                  </a>
                </th>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-3hphth-MuiTableCell-root"
                >
                  09/06/2023
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-3hphth-MuiTableCell-root"
                >
                  163:33
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
